models/model_upload/multimodal_models/vllm-miniCPM-2.6/1/model.py

```diff
@@ -0,0 +1,239 @@
+import base64
+import itertools
+import os
+import subprocess
+import sys
+import threading
+from typing import Iterator
+
+from clarifai.runners.models.model_runner import ModelRunner
+from clarifai.utils.logging import logger
+from clarifai_grpc.grpc.api import resources_pb2, service_pb2
```

Comment:
I think we should not have this level of abstraction here: 

from clarifai_grpc.grpc.api import resources_pb2, service_pb2
from clarifai_grpc.grpc.api.status import status_code_pb2

These create too much friction in a model building exercise. resources_pb2.Data, resources_pb2.Output are not straightforward to investigate for a model uploader

Reply:
service_pb2 as well. Why do I need to go through this? A good chunk of data scientists, research engineers will have no idea what to do with this.

Reply:
But this does not have to be this PR.

Reply:
@zeiler, what are your thoughts on this? I believe we discussed it a while ago, but I’m not sure if we reached a conclusion back then

Reply:
all the current examples are using our protos only at this point. I think we should publicly document our protos, especially Data. There could be alternatives to use the new numpy protos as well to have a more direct pythonic interface as well. 

Agreed not for this PR.

Reply:
Today I'm looking into `replicate` how they are dealing with `Inputs` and `Outputs` in `model.py`, I think `baseten` is also similar in dealing with  `Inputs` and `Outputs` though. 

**Input**
The predict() method takes an arbitrary list of named arguments, where each argument name must correspond to an [Input()](https://github.com/replicate/cog/blob/main/docs/python.md#inputkwargs) annotation.
Each parameter of the predict() method must be annotated with a type like str, int, float, bool, etc

```python
    def predict(self,
        prompt: str = "default prompt", # this is valid
        iterations: int                 # also valid
    ) -> str:
        # ...
```



**Output**
Cog predictors can return a simple data type like a string, number, float, or boolean. Use Python's -> <type> syntax to annotate the return type.

Here's an example of a predictor that returns a string:
```python
from cog import BasePredictor

class Predictor(BasePredictor):
    def predict(self) -> str:
        return "hello"
```
Returning an object
To return a complex object with multiple values, define an Output object with multiple fields to return from your predict() method:
```
from cog import BasePredictor, BaseModel, File

class Output(BaseModel):
    file: File
    text: str

class Predictor(BasePredictor):
    def predict(self) -> Output:
        return Output(text="hello", file=io.StringIO("hello"))
```

Returning a list output
The predict() method can return a list of any of the supported output types. Here's an example that outputs multiple files:
```
from cog import BasePredictor, Path

class Predictor(BasePredictor):
    def predict(self) -> list[Path]:
        predictions = ["foo", "bar", "baz"]
        output = []
        for i, prediction in enumerate(predictions):
            out_path = Path(f"/tmp/out-{i}.txt")
            with out_path.open("w") as f:
                f.write(prediction)
            output.append(out_path)
        return output
```

**Input and output types**
Each parameter of the predict() method must be annotated with a type. The method's return type must also be annotated. The supported types are:

str: a string
int: an integer
float: a floating point number
bool: a boolean
[cog.File](https://github.com/replicate/cog/blob/main/docs/python.md#file): a file-like object representing a file
[cog.Path](https://github.com/replicate/cog/blob/main/docs/python.md#path): a path to a file on disk
[cog.Secret](https://github.com/replicate/cog/blob/main/docs/python.md#secret): a string containing sensitive information

cc @zeiler @ackizilkale

Reply:
Why not design it the same way as the previous version?https://github.com/Clarifai/examples/blob/9263ad19d1ebe99dfdad5c1fae0bfc329ce77ed9/model_upload/text_to_text/hf-model/inference.py